File: Lab 5/Assembler/gasm.m
%%---------------------------------------------------------------
% Written by Gregory Leonberg for 332:493:02 "Embedded Systems Design"
% Fall 2017 Semester
% Converts custom GRISC ISA assembly into binary COE file
%%---------------------------------------------------------------

%%---------------------------------------------------------------
% format of input file is:
% [ignored lines]
% .data
% data segement
% .text
% text segement
%%---------------------------------------------------------------

%%---------------------------------------------------------------
% format of data segment is:
% [blank]                   OR
% // comment                OR
% * label: [type] [value] // comment
% types are str or num
% no spaces in string allowed
% [where comments are optional]
% [one operation or label per line]
% ignores whitespace
% assumes correct input, does not handle invalid assembly
% last line before .text MUST be a data statement
%%---------------------------------------------------------------

%%---------------------------------------------------------------
% format of text segment is:
% [blank]                   OR
% // comment                OR
% * opcode args // comment    OR
% * label: // comment
% [where comments are optional]
% [one operation or label per line]
% ignores whitespace
% assumes correct input, does not handle invalid assembly
%%---------------------------------------------------------------

% define opcode constants in associative array (dictonary)
ops = containers.Map;
ops('add') = '00000'; ops('sub') = '00001'; ops('sll') = '00010'; 
ops('srl') = '00011'; ops('sra') = '00100'; ops('and') = '00101';
ops('or') = '00110'; ops('xor') = '00111'; ops('slt') = '01000';
ops('sgt') = '01001'; ops('seq') = '01010'; ops('send') = '01011';
ops('recv') = '01100'; ops('jr') = '01101'; ops('wpix') = '01110';
ops('rpix') = '01111'; ops('beq') = '10000'; ops('bne') = '10001';
ops('ori') = '10010'; ops('lw') = '10011'; ops('sw') = '10100';
ops('j') = 	'11000'; ops('jal') = '11001'; ops('clrscr') = '11010';
ops('la') = '10010'; % pseudo instruction for ori with label as immediate

% define instruction type decoding in associative array (dictionary)
types = containers.Map;
types('00') = 'r';
types('01') = 'r';
types('10') = 'i';
types('11') = 'j';

% define register constants in associative array (dictionary)
regs = containers.Map;
regs('$zero') = '00000';
regs('$pc') = '00001';
regs('$ra') = '00010';
for i = 3:31    
    str = strcat('$r', num2str(i));
    regs(str) = dec2bin(i, 5);
end

% ask for path to input file
inpath = input('Please enter the path to the assembly file: ', 's');
cd(inpath);
inf = input('Please enter the input file name: ', 's');
inp = fopen(inf, 'r');

% handle output files
outpath = input('Please enter the name of the output text coe file: ', 's');
asm = fopen(outpath, 'w');
fprintf(asm, 'MEMORY_INITIALIZATION_RADIX=2;\n');
fprintf(asm, 'MEMORY_INITIALIZATION_VECTOR=\n');
outpathd = input('Please enter the name of the output data coe file: ', 's');
dat = fopen(outpathd, 'w');
fprintf(dat, 'MEMORY_INITIALIZATION_RADIX=2;\n');
fprintf(dat, 'MEMORY_INITIALIZATION_VECTOR=\n');

% cell array for lines of text
lines = cell(1);

% iterate across assembly code and strip comments and store into array
% strip lables and store addresses for labels
labels  = containers.Map();
dlabels = containers.Map();
dCounter = 0;
addressCounter = 0;

% advance to data segment
while(~strcmp(fgetl(inp), '.data'))
end

% parse data segment
templ = fgetl(inp);
while(~strcmp(templ, '.text'))
    astPos = find(templ == '*');
    if(~isempty(astPos))
        templ = templ(astPos+1:length(templ));
        commentsPos = strfind(templ, '//');
        if(size(commentsPos))
            templ = templ(1:commentsPos-1);
        end
        templ = strtrim(templ);
        
        % add current label to dict with current dCounter
        colonPos = find(templ == ':');
        dlabels(templ(1:colonPos-1)) = dec2bin(dCounter, 16);
        
        % convert data to groups of 16 bits and write to data coe
        args = textscan(templ, '%s', 'delimiter', ' ');
        dtype = args{1}{2};
        dval = args{1}{3};
        
        % convert data to binary and increment dcounter for next label
        if(strcmp(dtype, 'num'))
           binar = [dec2bin(str2num(dval), 16), ',\n'];
           dCounter = dCounter + 1;
        else
            binar = '';
            for i = 1:length(dval)
                binar = [binar, dec2bin(dval(i), 16), ',\n'];
                dCounter = dCounter + 1;
            end
                binar = [binar, '0000000000000000,\n'];
                dCounter = dCounter + 1;
        end
        
        % write binary to data coe file
        templ = fgetl(inp);
        if(strcmp(templ, '.text'))
            fprintf(dat, [binar(1:(length(binar)-3)), ';']);
        else
            fprintf(dat, binar);
        end
    
    else
        templ = fgetl(inp);
    end
end

% parse instruction segment
while ~feof(inp)
    
    % get line from file and strip comment if it exists and whitespace
    line = fgetl(inp);
    astPos = find(line == '*');
    if(~isempty(astPos))
        line = line(astPos+1:length(line));
        commentsPos = strfind(line, '//');
        if(size(commentsPos))
            line = line(1:commentsPos-1);
        end
        line = strtrim(line);

        % if not a label, increment address counter
        colonPos = find(line == ':');
        if(isempty(colonPos))
            addressCounter = addressCounter + 1;
            lines{addressCounter} = line;
        else
            labels(line(1:colonPos-1)) = dec2bin(addressCounter, 16);
        end
    end
end

% iterate across array and write convert each line, write to file
temp = size(lines);
numLines = temp(2);
for i = 1:numLines
    
    % split line into array by whitespace
    args = textscan(lines{i}, '%s', 'delimiter', ' ');
    opcode = args{1}{1};
    numArgs = size(args{1});
    numArgs = numArgs(1)-1;
    binop = ops(opcode);
    optype = types(binop(1:2));
    command = binop;
    
    % convert R-type
    if(optype == 'r')
        for j = 2:numArgs+1
            command = [command, regs(args{1}{j})]; 
        end
     
    elseif(optype == 'i')
        if(strcmp(opcode, 'la')) % la pseudo-instruction as ori
            command = [command, regs(args{1}{2}), regs('$zero'), dlabels(args{1}{3})];
        elseif(strcmp(opcode, 'lw') || strcmp(opcode, 'sw'))
            command = [command, regs(args{1}{2}), regs(args{1}{3}), dlabels(args{1}{4})];
        elseif(strcmp(opcode, 'beq') || strcmp(opcode, 'bne'))
            command = [command, regs(args{1}{2}), regs(args{1}{3}), labels(args{1}{4})];
        else
            temp =  dec2bin(str2double(args{1}{4}), 16);
            while(length(temp) < 16)
                temp = ['0', temp];
            end
            command = [command, regs(args{1}{2}), regs(args{1}{3}), temp];
        end
    
    % convert J-type
    else
        if(strcmp(opcode, 'j') || strcmp(opcode, 'jal'))
            command = [command, labels(args{1}{2})];
        else
            temp =  dec2bin(str2double(args{1}{4}), 16);
            while(length(temp) < 16)
                temp = ['0', temp];
            end
            command = [command, temp];
        end     
    end
    
    % pad command with blank 0s to right if smaller than 32 bits
    while(length(command) < 32)
                command = [command, '0'];
    end
    
    % write binary instruction to coe file
    fprintf(asm, command);
	if(i == numLines)
		fprintf(asm, ';');
	else
		fprintf(asm, ',\n');
    end
    
        
end

% close output file
fclose(inp);
fclose(asm);

File: Lab 5/AssemblerPython/gasm.py
##---------------------------------------------------------------
# Written by Gregory Leonberg for 332:493:02 "Embedded Systems Design"
# Fall 2017 Semester
# Converts custom GRISC ISA assembly into binary COE file
##---------------------------------------------------------------

##---------------------------------------------------------------
# format of input file is:
# [ignored lines]
# .data
# data segement
# .text
# text segement
##---------------------------------------------------------------

##---------------------------------------------------------------
# format of data segment is:
# [blank]                   OR
# // comment                OR
# label: [type] [value] // comment
# types are str or num
# [where comments are optional]
# [one operation or label per line]
# ignores whitespace
# assumes correct input, does not handle invalid assembly
##---------------------------------------------------------------

##---------------------------------------------------------------
# format of text segment is:
# [blank]                   OR
# // comment                OR
# opcode args // comment    OR
# label: // comment
# [where comments are optional]
# [one operation or label per line]
# ignores whitespace
# assumes correct input, does not handle invalid assembly
##---------------------------------------------------------------

from tkinter import filedialog
from tkinter import *

#################################################################

def assembler():
	
	# open input file
	sourceName = askOpen()
	source = open(sourceName, 'r')
	
	# open output files
	dataName = askSaveData()
	data = open(dataName, 'w')
	textName = askSaveText()
	text = open(textName, 'w')
	
	# go through input file, strip out all comments and spacing
	cleaned = list()
	for line in source:
		clean = line.strip().split("//")[0].strip()
		if clean:
			cleaned.append(clean)
	
	# go through cleaned up lines, split into data and text segments
	dStart = 1
	while cleaned[dStart-1] != '.data':
		dStart = dStart + 1
	tStart = dStart + 1
	while cleaned[tStart-1] != '.text':
		tStart = tStart + 1
	dataLines = cleaned[dStart:tStart-1]
	textLines = cleaned[tStart:]
	
	# build labels from data segment
	dLabels = dataLabels(dataLines)
	
	# convert data segment to binary COE file string array
	dataConvertedArr = buildDataCoe(dataLines)
	
	# build labels from instruction segment
	tLabels, textLines = textLabels(textLines)
	
	# convert text segment to binary COE file string array
	textConvertedArr = buildTextCoe(textLines, dLabels, tLabels)
	
	# write data segment to COE file
	for line in dataConvertedArr:
		data.write(line + '\n')
	
	# write text segment to COE file
	for line in textConvertedArr:
		text.write(line + '\n')
	
	return True

#################################################################

def dataLabels(dataLines):
	
	labels = {}
	counter = 0
	for line in dataLines:
		tag = line.split(':')[0]
		rest = line.split(':')[1]
		typ = rest.split()[0]
		val = rest.split(typ)[1].strip()
		
		if typ == 'str':
			val = val.split('"')[1]
			val = val.split('"')[0]
		
		labels[tag] = dec2bin(counter, 16)
		if typ == 'str':
			counter = counter + len(val) + 1
		else:
			counter = counter + 1
	return labels

def buildDataCoe(dataLines):
	
	lines = list()
	lines.append('MEMORY_INITIALIZATION_RADIX=2;')
	lines.append('MEMORY_INITIALIZATION_VECTOR=')
	
	for lineNum in range(len(dataLines)):
		line = dataLines[lineNum]
		tag = line.split(':')[0]
		rest = line.split(':')[1]
		typ = rest.split()[0]
		val = rest.split(typ)[1].strip()
		
		if typ == 'str':
			val = val.split('"')[1]
			val = val.split('"')[0]
			for char in val:
				lines.append(dec2bin(ord(char), 16) + ",")
			if lineNum < len(dataLines)-1:
				lines.append(dec2bin(0, 16) + ",")
			else:
				lines.append(dec2bin(0, 16) + ";")
		else:
			val = int(val)
			if lineNum < len(dataLines)-1:
				lines.append(dec2bin(val, 16) + ",")
			else:
				lines.append(dec2bin(val, 16) + ";")
	
	return lines

def textLabels(textLines):
	
	newLines = list()
	labels = {}
	counter = 0
	
	for line in textLines:
		if ":" in line:
			tag = line.split(':')[0]
			labels[tag] = dec2bin(counter, 16)
		else:
			newLines.append(line)
			counter = counter + 1
	
	return labels, newLines

def buildTextCoe(textLines, dLabels, tLabels):
	
	lines = list()
	lines.append('MEMORY_INITIALIZATION_RADIX=2;')
	lines.append('MEMORY_INITIALIZATION_VECTOR=')
	
	for lineNum in range(len(textLines)):
		
		ops, types, regs = buildDicts()
		
		args = textLines[lineNum].strip().split()
		opcode = args[0]
		numArgs = len(args)
		
		binop = ops[opcode]
		optype = types[binop[0:2]]
		command = binop
		
		# convert R-type
		if(optype == 'r'):
			for j in range(1, numArgs):
				command = command + regs[args[j]];
		
		# convert I-type
		elif(optype == 'i'):
			if(opcode == 'la'): # la pseudo-instruction as ori
				command = command + regs[args[1]] + regs['$zero'] + dLabels[args[2]]
			elif((opcode == 'lw') or (opcode == 'sw')):
				command = command + regs[args[1]] + regs[args[2]] + dLabels[args[3]]
			elif((opcode == 'beq') or (opcode == 'bne')):
				command = command + regs[args[1]] + regs[args[2]] + tLabels[args[3]]
			else:
				command = command + regs[args[1]] + regs[args[2]] + dec2bin(int(args[3]), 16)
		
		# convert J-type
		else:
			if((opcode == 'j') or (opcode == 'jal')):
				command = command + tLabels[args[1]]
			else:
				command = command + dec2bin(int(args[2]), 16)
		
		# pad command on the right to 32 bits
		command = command.ljust(32, '0')
		
		# write command to COE file list
		if lineNum < len(textLines)-1:
			lines.append(command + ",")
		else:
			lines.append(command + ";")
		
	return lines

def dec2bin(num, digits):
	ret = bin(num)
	ret = ret[2:]
	ret = ret.zfill(digits)
	return ret

def askOpen():
	root = Tk()
	root.filename = filedialog.askopenfilename(initialdir = "/",title = "Select source file",filetypes = (("assembly source","*.txt"),("all files","*.*")))
	return root.filename

def askSaveData():
	root = Tk()
	root.filename =  filedialog.asksaveasfilename(initialdir = "/",title = "Select data output file",filetypes = (("COE files","*.coe"),("all files","*.*")))
	return root.filename

def askSaveText():
	root = Tk()
	root.filename =  filedialog.asksaveasfilename(initialdir = "/",title = "Select text output file",filetypes = (("COE files","*.coe"),("all files","*.*")))
	return root.filename

def buildDicts():
	# define opcode constants in associative array (dictonary)
	ops = {};
	ops['add'] = '00000'
	ops['sub'] = '00001'
	ops['sll'] = '00010' 
	ops['srl'] = '00011'
	ops['sra'] = '00100'
	ops['and'] = '00101'
	ops['or'] = '00110'
	ops['xor'] = '00111'
	ops['slt'] = '01000'
	ops['sgt'] = '01001'
	ops['seq'] = '01010'
	ops['send'] = '01011'
	ops['recv'] = '01100'
	ops['jr'] = '01101'
	ops['wpix'] = '01110'
	ops['rpix'] = '01111'
	ops['beq'] = '10000'
	ops['bne'] = '10001'
	ops['ori'] = '10010'
	ops['lw'] = '10011'
	ops['sw'] = '10100'
	ops['j'] = 	'11000'
	ops['jal'] = '11001'
	ops['clrscr'] = '11010'
	ops['la'] = '10010' # pseudo instruction for ori with label as immediate
	
	# define instruction type decoding in associative array (dictionary)
	types = {};
	types['00'] = 'r';
	types['01'] = 'r';
	types['10'] = 'i';
	types['11'] = 'j';
	
	# define register constants in associative array (dictionary)
	regs = {};
	regs['$zero'] = '00000'
	regs['$pc'] = '00001'
	regs['$ra'] = '00010'
	for i in range(3, 32):
		st = '$r' + str(i)
		regs[st] = dec2bin(i, 5)
	
	return ops, types, regs

if __name__ == '__main__':
	success = assembler()
